lib/bq_stream/archive.rb

```diff
@@ -0,0 +1,239 @@
+      assign_back_date_id(table.constantize, back_date)
+
+      oldest_attr_recs = []
+      attrs.each do |a|
+        record = OldestRecord.where(table_name: table, attr: a.to_s).try(:first)
+        oldest_attr_recs << record if record
+      end
+      oldest_attr_recs.each { |rec| rec.update(archived: false) }
+      archive_table(table, oldest_attr_recs)
+
+      log(:info, "#{Time.now}: ***** End Partial Archive Process *****")
+    end
+
+    private
+
+    # Sets up BigQuery and means to write records to it  
+    def initialize_bq(override_dataset)
+      # Create BigQuery client connection
+      create_bq_writer(override_dataset)
+
+      # Create dataset if not present in BigQuery
+      @bq_writer.create_dataset(@bq_writer.dataset) unless @bq_writer.datasets_formatted.include?(@bq_writer.dataset)
+
+      # Create table in dataset if not present in BigQuery
+      unless @bq_writer.tables_formatted.include?(bq_table_name)
+        bq_table_schema = { table_name:   { type: 'STRING', mode: 'REQUIRED' },
+                            record_id:    { type: 'INTEGER', mode: 'REQUIRED' },
+                            attr:         { type: 'STRING', mode: 'NULLABLE' },
+                            new_value:    { type: 'STRING', mode: 'NULLABLE' },
+                            updated_at:   { type: 'TIMESTAMP', mode: 'REQUIRED' } }
+        @bq_writer.create_table(bq_table_name, bq_table_schema)
+      end
+    end
+
+    # Check to make sure all attributes that are in BigQuery and bq_attributes are represented in OldestRecord table
+    def update_oldest_records
+      old_records = @bq_writer.query('SELECT table_name, attr FROM '\
+                                      "[#{project_id}:#{@bq_writer.dataset}.#{bq_table_name}] "\
+                                      'GROUP BY table_name, attr')
+      if old_records['rows']
+        old_records['rows'].each do |r|
+          table = r['f'][0]['v']
+          trait = r['f'][1]['v']
+          if !trait.nil? && bq_attributes[table] && bq_attributes[table].include?(trait.to_sym) 
+            OldestRecord.find_or_create_by(table_name: table, attr: trait)
+          else
+            if !trait.nil? && bq_attributes[table]
+              log(:info, "#{Time.now}: No longer tracking: #{table}: #{trait}")
+            end
+          end
+        end
+      end
+    end
+
+    # Add and Remove bq_attributes based on current bq_attributes
+    def verify_oldest_records
+      log(:info, "#{Time.now}: ***** Start Verifying Oldest Records *****")
+      current_deploy =
+        if (`cat #{File.expand_path ''}/REVISION`).blank?
+          'None'
+        else
+          `cat #{File.expand_path ''}/REVISION`
+        end
+      bq_attributes.each do |k, v|
+        # add any records to oldest_records that are new (Or more simply make sure that that there is a record using find_by_or_create)
+        v.each do |bqa|
+          OldestRecord.find_or_create_by(table_name: k, attr: bqa)
+        end
```

Comment:
maybe use more descriptive variable names?

```ruby
bq_attributes.each do |model, attrs|
  attrs.each do |attr|
    OldestRecord.find_or_create_by(table_name: model, attr: attr)
  end

  OldestRecord.where(table_name: model).each do |record|
    record.destroy unless attrs.include?(record.attr.to_sym)
  end
end
```

This way it's obvious what the data that you're manipulating is.